### PR TITLE
Dispatch - Talend Java Process - graceful termination of long running jobs

### PR DESCRIPTION
**Pull request by igorbeslic**
Hi Shuyang please check this one. Thank you very much!

### CODE REVIEW

File: modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcessCallable.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		try {
+			Class<?> talendJobClass = classLoader.loadClass(
+				_mainMethodClassFQN);
+
+			Method talendJobClassMainMethod = talendJobClass.getMethod(
+				"main", String[].class);
+
+			talendJobClassMainMethod.setAccessible(true);
+
+			talendJobClassMainMethod.invoke(
```
Comment (igorbeslic) [inline modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcessCallable.java:50]: Hi @shuyangzhou Talend terminates JVM in its own main method - what is quite OK and LocalProcessExecutor handles it correctly.

I wonder what is your opinion on termination of long running Talend job. Talend does not offer graceful shutdown method and I would like to provide shutdown hook in lambda form where I would like to invoke System.exit(0) from TalendProcessCollable.
In com.liferay.portal.search.elasticsearch7.internal.sidecar.ElasticsearchServerUtil that approach is used.

**Review by igorbeslic — COMMENTED** [review]
accidentally clicked wrong button

File: modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcess.java [hunk 2]
```diff
@@ -112,6 +135,19 @@ public Builder companyId(long companyId) {
 		}
 
 		public Builder contextParam(String name, String value) {
+			if (name.startsWith("-D") || name.startsWith("-X")) {
```
Comment (igorbeslic) [inline modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcess.java:138]: @shuyangzhou do we have any Util I could use to handle JVM options? User can pass options via UI and I would like to validate it before passing to sub-process.

File: modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/executor/TalendDispatchTaskExecutor.java [hunk 3]
```diff
@@ -119,6 +169,46 @@ protected TalendArchive fetchTalendArchive(long dispatchTriggerId)
...
+
+			return sb.toString();
+		}
+		catch (IOException ioException) {
+			throw new RuntimeException(
+				"Unable to iterate " + dirPath, ioException);
+		}
+	}
+
+	private String _getBootstrapClassPath() {
```
Comment (igorbeslic) [inline modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/executor/TalendDispatchTaskExecutor.java:198]: Hi @shuyangzhou you instructed me to check/use com.liferay.portal.util.PortalClassPathUtil but it is in portal-impl and it is not available outside.
Do you know if there is another util that can help me to reduce code in this class?? I copy pasted code from Sidecar and I would rather like one specialized Util that can benefit to Sidecar, Dispatch and other future modules.

File: modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcessCallable.java (hunk 1 above)
Comment (shuyangzhou) [reply modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcessCallable.java:50]: For force termination, you need to hold on the "ProcessChannel.getProcessNoticeableFuture()" to call cancel on it. The ES sidecar solution is to workaround the complex internal details from ES, you don't need that.

File: modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcess.java (hunk 2 above)
Comment (shuyangzhou) [reply modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/process/TalendProcess.java:138]: We have no such thing, you have to collect your own jvm args.

File: modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/executor/TalendDispatchTaskExecutor.java (hunk 3 above)
Comment (shuyangzhou) [reply modules/apps/dispatch/dispatch-talend-web/src/main/java/com/liferay/dispatch/talend/web/internal/executor/TalendDispatchTaskExecutor.java:198]: You can use PortalClassPathUtil just fine, portal-impl utils are available from modules.

**Comment by shuyangzhou** [comment]
See the answer above, when you send back the final version to review, make sure to add an integration test for this. I have no idea how to use this, I will need to reply on the integration test to try and review the code.

**Pull request closed without merging**